### ISSUE TITLE
pkg/util/ratelimiter	race with go1.5

### ISSUE THREAD

**Issue by soltysh** [openshift/origin#7785]
When running test I'm getting following race warning:
```
==================
WARNING: DATA RACE
Read by goroutine 16:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction.func3()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:51 +0x87

Previous write by goroutine 14:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:46 +0x3df
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc

Goroutine 16 (running) created at:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:53 +0x3b5
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc

Goroutine 14 (running) created at:
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:494 +0xe4
  main.main()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_testmain.go:100 +0x384
==================
==================
WARNING: DATA RACE
Write by goroutine 14:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:35 +0x1af
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc

Previous read by goroutine 31:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction.func3()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:48 +0x37

Goroutine 14 (running) created at:
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:494 +0xe4
  main.main()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_testmain.go:100 +0x384

Goroutine 31 (finished) created at:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:53 +0x3b5
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc
==================
==================
WARNING: DATA RACE
Write by goroutine 21:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction.func2()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:38 +0x54
  github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).handleOne()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_obj_test/ratelimiter.go:56 +0x9a
  github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).RunUntil.func1()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_obj_test/ratelimiter.go:48 +0xaf
  k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:66 +0x58
  k8s.io/kubernetes/pkg/util/wait.JitterUntil()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:67 +0x7d
  k8s.io/kubernetes/pkg/util/wait.Until()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47 +0x4b

Previous read by goroutine 14:
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:58 +0x4b2
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc

Goroutine 21 (running) created at:
  github.com/openshift/origin/pkg/util/ratelimiter.(*RateLimitedFunction).RunUntil()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_obj_test/ratelimiter.go:48 +0xd2
  github.com/openshift/origin/pkg/util/ratelimiter.TestRateLimitedFunction()
      /home/maszulik/workspace/origin/src/github.com/openshift/origin/pkg/util/ratelimiter/ratelimiter_test.go:44 +0x319
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:456 +0xdc

Goroutine 14 (running) created at:
  testing.RunTests()
      /usr/lib/golang/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:494 +0xe4
  main.main()
      github.com/openshift/origin/pkg/util/ratelimiter/_test/_testmain.go:100 +0x384
==================
PASS
coverage: 91.7% of statements
Found 3 data race(s)
FAIL	github.com/openshift/origin/pkg/util/ratelimiter	15.027s
```

**Comment by soltysh** [comment]
@ramr there are a few problems with it:
1. you should pass the variables to the goroutine instead of using the ones from for loop (see this [reference](http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#closure_for_it_vars))
2. you'll have to introduce some locks to the counter value

**Comment by ramr** [comment]
@soltysh sorry had a long day and just catching up on this. Thx for the PR. The only fixes probably  needed for go 1.5 really are the lock around the counter (not do dirty reads) and the close. The index variable passed to Invoke method is not really required for this test - we are just checking that its called - that's all - the resource its called [the index variable in this case] on is not important. But it doesn't hurt to do your changes either.
